src/db/group.rs

```diff
@@ -148,6 +157,156 @@ impl Group {
+                        *e = entry.clone();
+                    }
+                }
+            }
+        }
+    }
+
+    pub fn find_entry_by_uuid(&self, id: Uuid) -> Option<&Entry> {
+        for node in &self.children {
+            match node {
+                Node::Group(g) => {
+                    if let Some(e) = g.find_entry_by_uuid(id) {
+                        return Some(e);
+                    }
+                }
+                Node::Entry(e) => {
+                    if e.uuid == id {
+                        return Some(e);
+                    }
+                }
+            }
+        }
+        None
+    }
+
+    pub(crate) fn add_entry(&mut self, entry: Entry, location: &NodeLocation) {
+        if location.len() == 0 {
+            panic!("TODO handle this with a Response.");
+        }
+
+        let mut remaining_location = location.clone();
+        remaining_location.remove(0);
+
+        if remaining_location.len() == 0 {
+            self.children.push(Node::Entry(entry.clone()));
+            return;
+        }
+
+        let next_location = &remaining_location[0];
+
+        println!(
+            "Searching for group {} {:?}",
+            next_location.name, next_location.uuid
+        );
+        for node in &mut self.children {
+            if let Node::Group(g) = node {
+                if g.uuid != next_location.uuid {
+                    continue;
+                }
+                g.add_entry(entry, &remaining_location);
+                return;
+            }
+        }
+
+        // The group was not found, so we create it.
+        let mut new_group = Group {
+            name: next_location.name.clone(),
+            uuid: next_location.uuid.clone(),
+            ..Default::default()
+        };
+        new_group.add_entry(entry, &remaining_location);
+        self.children.push(Node::Group(new_group));
+    }
+
+    /// Merge this group with another group
+    pub fn merge(&mut self, other: &Group) {
+        for (entry, entry_location) in other.get_all_entries(&vec![]) {
+            if let Some(existing_entry) = self.find_entry_by_uuid(entry.uuid) {
+                if existing_entry == entry {
+                    continue;
+                }
+                // TODO relocate the existing entry if necessary
+                let merged_entry = existing_entry.merge(entry);
+                self.replace_entry(&merged_entry);
+            } else {
+                println!("Adding entry {} at {:?}", entry.uuid, &entry_location);
+                self.add_entry(entry.clone(), &entry_location);
+                // TODO should we update the time info for the entry?
+            }
+        }
+
+        // TODO update locations
+        // TODO handle deleted objects
+    }
+
+    // Recursively get all the entries in the group, along with their
+    // location.
+    pub(crate) fn get_all_entries(
+        &self,
+        current_location: &NodeLocation,
+    ) -> Vec<(&Entry, NodeLocation)> {
```

Comment:
This should be replaced with a recursive iterator implementation as suggested above.